### PR TITLE
Remap Mouse and Key Buttons

### LINKED ISSUE #2361: Option to remap mouse buttons, not just keyboard keys
I would like to have the ability to remap my mouse buttons freely. Like middleclick = placing nodes or rightlick = jump for example.

Would be a useful feature for lefthanded players, too.

See MT forum thread:
https://forum.minetest.net/viewtopic.php?f=5&t=11255

### PR DESCRIPTION
**Pull request by fz72**
fixes #2361 

Remap mouse buttons in keychange menu.
Mouse Button Events are now handled like key button events.

### CODE REVIEW

**Comment by kilbith** [comment]
#2361 request for **remap** the mouse buttons set, not only inverse two of them. Therefore it's a partial fix.

**Comment by fz72** [comment]
I will rework this, cause the actual solution is ugly.

**Comment by ShadowNinja** [comment]
@fz72 No, you could, eg, remap digging to the `g` key, or map forward/backward movement to the left and right mouse buttons.  There's also a third mouse button (the scroll wheel has a button in it).

**Comment by Rochambeau** [comment]
Actually my mouse has five buttons. :-)

My preffered setting would be:
-left: digging
-middle: place
-right: jump
-thumb: inventory
-ringfinger: walk

But left, right and middle would be sufficient for now, if it's too complicated.

**Comment by fz72** [comment]
What do you guys think about another lib for key and moues event. This one supports all mouse buttons. http://irrlicht3d.org/wiki/index.php?n=Main.IntegratingOISWithIrrlicht

File: src/client/inputhandler.h [hunk 1]
```diff
@@ -60,44 +72,87 @@ class MyEventReceiver : public IEventReceiver
...
 		}
+
 		if (event.EventType == irr::EET_LOG_TEXT_EVENT) {
 			dstream << std::string("Irrlicht log: ") + std::string(event.LogEvent.Text)
 			        << std::endl;
 			return true;
 		}
+		if (event.EventType == EET_KEY_INPUT_EVENT) {
+
+		}
```
Comment (nerzhul) [inline src/client/inputhandler.h]: Why is this empty ?

Comment (fz72) [inline src/client/inputhandler.h]: EET_KEY_INPUT_EVENT is handled at the beginning of this method. I will remove it.

File: src/client/inputhandler.h [hunk 2]
```diff
@@ -60,32 +72,62 @@ class MyEventReceiver : public IEventReceiver
...
+			// left
+			if (event.MouseInput.Event == EMIE_LMOUSE_PRESSED_DOWN) {
+			     	if (keyIsDown["KEY_LBUTTON"]) {
+					keyPressed.set("KEY_LBUTTON");
+			     	}
+				keyDown.set("KEY_LBUTTON");
+				keyIsDown.set("KEY_LBUTTON");
+				keyWasDown.set("KEY_LBUTTON");
+		  	}
+		  	if (event.MouseInput.Event == EMIE_LMOUSE_LEFT_UP) {
```
Comment (Zeno-) [inline src/client/inputhandler.h]: I know you're using the same "pattern" as the code you replaced but this (and the ones below) should really be `else if`

File: src/client/inputhandler.h [hunk 3]
```diff
@@ -60,32 +72,62 @@ class MyEventReceiver : public IEventReceiver
...
+				keyWasDown.set("KEY_MBUTTON");
+		  	}
+		  	if (event.MouseInput.Event == EMIE_MMOUSE_LEFT_UP) {
+		     		keyDown.unset("KEY_MBUTTON");
+		     		keyIsDown.set("KEY_MBUTTON");
+	     			keyPressed.unset("KEY_MBUTTON");
+				keyReleased.set("KEY_MBUTTON");	
+		  	}
+
+			if (noMenuActive()) {
```
Comment (Zeno-) [inline src/client/inputhandler.h]: Combine with condition below... i.e.

```
else if (event.MouseInput.Event == EMIE_MOUSE_WHEEL && noMenuActive()) {
    mouse_wheel += event.MouseInput.Wheel;
```

File: src/client/inputhandler.h [hunk 4]
```diff
@@ -269,72 +293,97 @@ class RandomInputHandler : public InputHandler
...
 		rightreleased = false;
+		middlereleased = false;
 		keydown.clear();
 	}
 	virtual bool isKeyDown(const KeyPress &keyCode)
 	{
+		if (keyCode == getKeySetting("keymap_dig")) {
+			return leftclicked;
+		}
+		if (keyCode == getKeySetting("keymap_place")) {
```
Comment (Zeno-) [inline src/client/inputhandler.h]: else if
(I don't see how keyCode can have more than one value at once. I guess getKeySetting("keymap_dig") and getKeySetting("keymap_place") might be able to map to the same keyCode but surely that would be a bug)

File: src/game.cpp [hunk 5]
```diff
@@ -2958,23 +2965,23 @@ void Game::updatePlayerControl(const CameraOrientation &cam)
...
-		( (u32)(input->isKeyDown(keycache.key[KeyCache::KEYMAP_ID_FORWARD])  & 0x1) << 0) |
-		( (u32)(input->isKeyDown(keycache.key[KeyCache::KEYMAP_ID_BACKWARD]) & 0x1) << 1) |
-		( (u32)(input->isKeyDown(keycache.key[KeyCache::KEYMAP_ID_LEFT])     & 0x1) << 2) |
-		( (u32)(input->isKeyDown(keycache.key[KeyCache::KEYMAP_ID_RIGHT])    & 0x1) << 3) |
-		( (u32)(input->isKeyDown(keycache.key[KeyCache::KEYMAP_ID_JUMP])     & 0x1) << 4) |
-		( (u32)(input->isKeyDown(keycache.key[KeyCache::KEYMAP_ID_SPECIAL1]) & 0x1) << 5) |
-		( (u32)(input->isKeyDown(keycache.key[KeyCache::KEYMAP_ID_SNEAK])    & 0x1) << 6) |
-		( (u32)(input->getLeftState()                                        & 0x1) << 7) |
-		( (u32)(input->getRightState()                                       & 0x1) << 8
+		( (u32)(input->isKeyDown(keycache.key[KeyCache::KEYMAP_ID_FORWARD])  	& 0x1) << 0) |
```
Comment (Zeno-) [inline src/game.cpp]: I'm not sure we align with spaces + tabs (only spaces I think).
@ShadowNinja ?

File: src/guiKeyChangeMenu.cpp [hunk 6]
```diff
@@ -93,13 +95,11 @@ void GUIKeyChangeMenu::removeChildren()
 	const core::list<gui::IGUIElement*> &children = getChildren();
 	core::list<gui::IGUIElement*> children_copy;
 	for (core::list<gui::IGUIElement*>::ConstIterator i = children.begin(); i
-		 != children.end(); i++)
-	{
+		 != children.end(); i++) {
```
Comment (Zeno-) [inline src/guiKeyChangeMenu.cpp]: While you're fixing the code style also put the whole condition on its own line ;)

File: src/guiKeyChangeMenu.cpp [hunk 7]
```diff
@@ -268,29 +265,45 @@ bool GUIKeyChangeMenu::resetMenu()
...
 				event.KeyInput.Key == irr::KEY_RSHIFT))
 			shift_went_down = true;
-
+	}
+	else if (event.EventType == EET_MOUSE_INPUT_EVENT && activeKey >= 0) {
+		if (event.MouseInput.Event == EMIE_LMOUSE_PRESSED_DOWN) {
+			kp = KeyPress("KEY_LBUTTON");
+			new_key = true;
+		}
+		if (event.MouseInput.Event == EMIE_RMOUSE_PRESSED_DOWN) {
```
Comment (Zeno-) [inline src/guiKeyChangeMenu.cpp]: else if?

Comment (fz72) [inline src/guiKeyChangeMenu.cpp]: I will fix this later ;)

**Comment by Zeno-** [comment]
Apart from some minor style issues (else and else if should be on the same line as the closing brace of the previous compound statement in connection.h) I approve this

:+1:

File: src/client/inputhandler.h [hunk 8]
```diff
@@ -60,32 +71,59 @@ class MyEventReceiver : public IEventReceiver
...
-				if (event.MouseInput.Event == EMIE_RMOUSE_LEFT_UP) {
-					rightreleased = true;
-				}
+			// handle mouse buttons
+			// left
+			if (event.MouseInput.Event == EMIE_LMOUSE_PRESSED_DOWN) {
+			     	if (keyIsDown["KEY_LBUTTON"]) {
+					keyPressed.set("KEY_LBUTTON");
+			     	}
+				keyDown.set("KEY_LBUTTON");
```
Comment (Zeno-) [inline src/client/inputhandler.h]: Is stuff like this really necessary?

Comment (fz72) [inline src/client/inputhandler.h]: Yes, I think so. To handle the mouse buttons I need to write the states of them into keyIsDown keyPressed and so on.

Comment (fz72) [inline src/client/inputhandler.h]: @Zeno- I think we should let so make another pull request with a rework of the KeyList Class.

**Comment by paramat** [comment]
No activity for a year.

**Comment by Wuzzy2** [comment]
I hope this feature gets finished/implemented before it dies off. :-(

File: src/game.cpp [hunk 9]
```diff
@@ -1271,6 +1275,9 @@ struct KeyCache {
 
 void KeyCache::populate()
 {
+	key[KEYMAP_ID_DIG]      = getKeySetting("keymap_dig");
+	key[KEYMAP_ID_PLACE]      = getKeySetting("keymap_place");
```
Comment (SmallJoker) [inline src/game.cpp]: The equal signs should be on the same level, just like below.

File: src/client/inputhandler.h [hunk 10]
```diff
@@ -419,10 +465,13 @@ class RandomInputHandler : public InputHandler
 	v2s32 mousespeed;
 	bool leftdown;
 	bool rightdown;
+	bool middledown;
 	bool leftclicked;
 	bool rightclicked;
+	bool middleclicked;
 	bool leftreleased;
 	bool rightreleased;
+	bool middlereleased;
```
Comment (SmallJoker) [inline src/client/inputhandler.h]: `middle*` are currently not used - what will happen with them?

**Comment by SmallJoker** [comment]
Developer gone - closed due inactivity.
The issue will still contain a reference to the pull in case someone would like to adapt it.
